project_timesheet_holidays_contract/models/hr_employee.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2024 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from datetime import datetime
+
+from odoo import _, models
+from odoo.exceptions import ValidationError
+
+
+class HrEmployee(models.Model):
+    _inherit = "hr.employee"
+
+    def _get_calendar(self, date_from=None):
+        res = super()._get_calendar(date_from)
+        if not date_from:
+            return res
+
+        contracts = (
+            self.sudo()
+            .with_context(active_test=True)
+            .env["hr.contract"]
+            .search([("employee_id", "=", self.id)])
+        )
+
+        if date_from:
+            date_from = (
+                date_from if not isinstance(date_from, datetime) else date_from.date()
+            )
+            contracts = contracts.filtered(lambda c: c.date_start <= date_from)
+
+        date_to = self.env.context.get("date_to")
+        if date_to:
+            date_to = date_to if not isinstance(date_to, datetime) else date_to.date()
+            contracts = contracts.filtered(
+                lambda c: not c.date_end or c.date_end >= date_to
+            )
```

Comment:
Filtering after searching is a bad practice perf-wise, and should be avoided unless necessary (for example, filtering the recordset according to a non-searchable field is ok, since you cannot do it otherwise).
Please handle both `date_from` and `date_to` variables before the `search()` and implement its domain according to those parameters, instead of searching and filtering.